File: QytetetRuby/lib/qytetet.rb
#encoding: UTF-8

module ModeloQytetet
  require 'singleton'
  require_relative 'tablero'
  require_relative 'jugador'
  require_relative 'dado'
  class Qytetet
    include Singleton
    MAX_JUGADORES = 4
    MAX_CARTAS = 10
    PRECIO_LIBERTAD = 200
    SALDO_SALIDA = 1000
    attr_reader :carta_actual, :jugador_actual, :tablero, :jugadores, :mazo

    def initialize
      @carta_actual = nil
      @mazo = Array.new
      @jugador_actual = nil
      @jugadores = Array.new
      @dado = Dado.instance
      @tablero = nil
    end

    def aplicar_sorpresa
      tiene_propietario = false
      case @carta_actual.tipo
      when TipoSorpresa::PAGARCOBRAR
        @jugador_actual.modificar_saldo(@carta_actual.valor)
      when TipoSorpresa::IRACASILLA
        if @tablero.es_casilla_carcel(@carta_actual.valor)
          encarcelar_jugador
        else
          nueva_casilla = @tablero.obtener_casilla_numero(@carta_actual.valor)
          tiene_propietario = @jugador_actual.actualizar_posicion(nueva_casilla)
        end
      when TipoSorpresa::PORCASAHOTEL
        @jugador_actual.pagar_cobrar_por_casa_y_hotel(@carta_actual.valor)
      when TipoSorpresa::PORJUGADOR
        @jugadores.reject { |j| j == @jugador_actual }.each do |j|
          j.modificar_saldo(@carta_actual.valor)
          @jugador_actual.modificar_saldo(-@carta_actual.valor)
        end
      end

      if @carta_actual.tipo == TipoSorpresa::SALIRCARCEL
        @jugador_actual.carta_libertad = @carta_actual
      else
        @mazo << @carta_actual
      end
      #EXAMEN-INICIO
      if @carta_actual.tipo TipoSorpresa::LOTERIA
        @jugador_actual.carta_loteria = @carta_actual
      else
        @mazo << @carta_actual
      end
      #EXAMEN-FIN
      tiene_propietario
    end

    def cancelar_hipoteca(casilla)
      return false if !casilla.esta_hipotecada
      @jugador_actual.modificar_saldo(-casilla.cancelar_hipoteca)
      true
    end

    def comprar_titulo_propiedad
      @jugador_actual.comprar_titulo
    end

    def edificar_casa(casilla)
      puedo_edificar = false
      if casilla.soy_edificable && casilla.se_puede_edificar_casa
        puedo_edificar = @jugador_actual.puedo_edificar_casa(casilla)
        if puedo_edificar
          coste_edificar_casa = casilla.edificar_casa
          @jugador_actual.modificar_saldo(-coste_edificar_casa)
        end
      end
      puedo_edificar
    end

    def edificar_hotel(casilla)
      puedo_edificar = false
      if casilla.soy_edificable && casilla.se_puede_edificar_hotel
        puedo_edificar = @jugador_actual.puedo_edificar_hotel(casilla)
        if puedo_edificar
          coste_edificar_hotel = casilla.edificar_hotel
          @jugador_actual.modificar_saldo(-coste_edificar_hotel)
        end
      end
      puedo_edificar
    end

    def hipotecar_propiedad(casilla)
      return if casilla.soy_edificable && !casilla.esta_hipotecada &&
        @jugador_actual.puedo_hipotecar(casilla)
      @jugador_actual.modificar_saldo(casilla.hipotecar)
    end

    def intentar_salir_carcel(metodo)
      libre = false
      if metodo == MetodoSalirCarcel::TIRANDODADO
        valor_dado = @dado.tirar
        libre = valor_dado > 4
      elsif metodo == MetodoSalirCarcel::PAGANDOLIBERTAD
        libre = @jugador_actual.pagar_libertad(PRECIO_LIBERTAD)
      end
      @jugador_actual.encarcelado = false if libre
      libre
    end

    def jugar
      valor_dado        = @dado.tirar
      casilla_posicion  = @jugador_actual.casilla_actual
      nueva_casilla     = @tablero.obtener_nueva_casilla(casilla_posicion, 2)
      tengo_propietario = @jugador_actual.actualizar_posicion(nueva_casilla)

      if !nueva_casilla.soy_edificable
        if nueva_casilla.tipo == TipoCasilla::JUEZ
          encarcelar_jugador
        elsif nueva_casilla.tipo == TipoCasilla::SORPRESA
          @carta_actual = @mazo.shift # Saca la sorpresa
        end
      end

      { dado: valor_dado, tiene_propietario: tengo_propietario }
    end

    def obtener_ranking
      r = @jugadores.sort_by(&:obtener_capital)
      .map { |j| [j.nombre, j.obtener_capital] }
      Hash[r]
    end

    def propiedades_hipotecadas_jugador(hipotecadas)
      @jugador_actual.obtener_propiedades_hipotecadas(hipotecadas).map(&:casilla)
    end

    def siguiente_jugador
      @jugador_actual = @jugadores[
        (@jugadores.find_index(@jugador_actual) + 1) % @jugadores.size]
    end

    def vender_propiedad(casilla)
      puedo_vender = false
      if casilla.soy_edificable &&
          @jugador_actual.puedo_vender_propiedad(casilla)
        @jugador_actual.vender_propiedad(casilla)
        puedo_vender = true
      end
      puedo_vender
    end

    def encarcelar_jugador
      if @jugador_actual.tengo_carta_libertad
        carta = @jugador_actual.devolver_carta_libertad
        @mazo << carta
      else
        @jugador_actual.ir_a_carcel(@tablero.carcel)
      end
    end

    def siguiente_jugador
      num_jug_sig = (@jugadores.index(@jugador_actual) + 1) % @jugadores.size
      @jugador_actual = @jugadores.at(num_jug_sig)
    end

    def inicializar_cartas_sorpresa
      @mazo<< Sorpresa.new("Se han limpiado tus delitos de la base de datos de la policia. Sales de la cárcel", 0, TipoSorpresa::SALIRCARCEL)
      @mazo<< Sorpresa.new("Te hemos pillado hackeando los servidores de la UGR, vas directamente a la carcel",
        @tablero.carcel.numero_casilla, TipoSorpresa::IRACASILLA )
      @mazo<< Sorpresa.new("Has ganado un viaje un viaje a Las Vegas, pero lo vendes porque prefieres seguir programando",
        800, TipoSorpresa::PAGARCOBRAR )
      @mazo<< Sorpresa.new("Decides ir de compras, vas en metro a Recogidas", 19, TipoSorpresa::IRACASILLA)
      @mazo<< Sorpresa.new("Un huracán ha arrasado con todos tus hoteles, pagas 300 por cada casa y hotel.", 100, TipoSorpresa::PORCASAHOTEL)
      @mazo<< Sorpresa.new("Un magnate del petroleo ha decidido invertir en tus propiedades,\
        obtienes 400 por cada casa y hotel", 400, TipoSorpresa::PORCASAHOTEL)
      @mazo<< Sorpresa.new("Decides regalar un año de suscripción de Netflix a cada jugador, \
        pagas 150 por cada jugador", 150, TipoSorpresa::PORJUGADOR)
      @mazo<< Sorpresa.new("Has aprobado la carrera, cada jugador te da 200 por \
        tu gran hazaña", 200, TipoSorpresa::PORJUGADOR)
      @mazo<< Sorpresa.new("Tienes ganas de salir de fiesta, vas a Pedro Antonio", 8, TipoSorpresa::IRACASILLA)
      @mazo<< Sorpresa.new("Anoche te pasaste con la juerga, te has dejado 750 en una noche", 750, TipoSorpresa::PAGARCOBRAR)
      @mazo<< Sorpresa.new("Te ha tocado la lotería, disfruta de Coco Bomgo", 1000000, TipoSorpresa::LOTERIA)
      @mazo<< Sorpresa.new("Te ha llegado un sobre con una tarjeta negra, eres un nuevo especulador", 3000, TipoSorpresa::CONVERTIRME)
      @mazo<< Sorpresa.new("Bonita cuenta en Suiza, ahora eres un especulador", 5000, TipoSorpresa::CONVERTIRME)
    end

    def inicializar_jugadores(nombres)
      nombres.each { |nombre| @jugadores<<Jugador.new(nombre)  }
    end

    def inicializar_tablero
      @tablero = Tablero.new
    end

    def salida_jugadores
      for jugador in @jugadores
        jugador.casilla_actual = @tablero.obtener_casilla_numero(0)
        jugador.modificar_saldo(7500)
      end
      @jugador_actual = @jugadores.sample
    end

    def inicializar_juego (nombres)
      inicializar_jugadores(nombres)
      inicializar_tablero
      inicializar_cartas_sorpresa
      salida_jugadores
    end

    #EXAMEN-INICIO
    def irme_a_vivir_a_cancun
      return false unless @jugador_actual.tengo_carta_loteria
      ganancia_ventas = @jugador_actual.vender_todo
      ganancia_hipotecar = @jugador_actual.hipotecar_todo

      propiedades = @jugador_actual.obtener_propiedades_hipotecadas(false)
      if ganancia_ventas > ganancia_hipotecar
        propiedades.each { |prop|
          @jugador_actual.vender_propiedad(prop.casilla)}
      else
        propiedades.each { |prop|
          hipotecar_propiedad(prop.casilla)}
      end

      cantidad = @jugador_actual.carta_loteria.valor
      @jugador_actual.modificar_saldo(cantidad)
      @jugador_actual.devolver_carta_loteria
      true
    end

    private :encarcelar_jugador, :inicializar_cartas_sorpresa,
      :inicializar_jugadores, :inicializar_tablero,
      :salida_jugadores
  end
end

File: QytetetJava/src/modeloqytetet/Qytetet.java
package modeloqytetet;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

public class Qytetet {

    public final int MAX_JUGADORES;
    public final int MAX_CARTAS;
    public final int MAX_CASILLAS;
    public final int PRECIO_LIBERTAD;
    public final int SALDO_SALIDA;
    private static final Qytetet instance = new Qytetet();
    private Dado dado;
    private Sorpresa cartaActual;
    private ArrayList<Sorpresa> mazo;
    private ArrayList<Jugador> jugadores;
    private Jugador jugadorActual;
    private Tablero tablero;
    private int valorDado;

    private Qytetet() {
        this.MAX_JUGADORES = 4;
        this.MAX_CARTAS = 10;
        this.MAX_CASILLAS = 20;
        this.SALDO_SALIDA = 1000;
        this.PRECIO_LIBERTAD = 200;
        this.mazo = new ArrayList();
        this.jugadores = new ArrayList();
        dado = Dado.getInstancia();
        cartaActual = null;
        valorDado = 0;
    }

    public Tablero getTablero() {
        return this.tablero;
    }

    public Sorpresa getCartaActual() {
        return cartaActual;
    }

    public Jugador getJugadorActual() {
        return jugadorActual;
    }

    public int getValorDado() {
        return valorDado;
    }

    private void inicializarCartasSorpresa() {
        mazo.add(new Sorpresa("Bonita cuenta en Suiza, ahora eres un especulador", 5000, TipoSorpresa.CONVERTIRME));
        mazo.add(new Sorpresa("Te hemos pillado hackeando los servidores de la UGR, vas directamente a la carcel",
                tablero.getCarcel().getNumeroCasilla(), TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa("Decides ir de compras, vas en metro a Recogidas", 19, TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa("Se han limpiado tus delitos de la base de datos de la policia. Sales de la cárcel",
                0, TipoSorpresa.SALIRCARCEL));
        mazo.add(new Sorpresa("Has ganado un viaje un viaje a Las Vegas, pero lo vendes porque prefieres seguir programando",
                800, TipoSorpresa.PAGARCOBRAR));
        mazo.add(new Sorpresa("Un huracán ha arrasado con todos tus hoteles, pagas 300 por cada casa y hotel.", 100, TipoSorpresa.PORCASAHOTEL));
        mazo.add(new Sorpresa("Un magnate del petroleo ha decidido invertir en tus propiedades, "
                + "obtienes 400 por cada casa y hotel", 400, TipoSorpresa.PORCASAHOTEL));
        mazo.add(new Sorpresa("Decides regalar un año de suscripción de Netflix a cada jugador, "
                + "pagas 150 por cada jugador", 150, TipoSorpresa.PORJUGADOR));
        mazo.add(new Sorpresa("Has aprobado la carrera, cada jugador te da 200 por"
                + "tu gran hazaña", -200, TipoSorpresa.PORJUGADOR));
        mazo.add(new Sorpresa("Tienes ganas de salir de fiesta, vas a Pedro Antonio", 8, TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa("Anoche te pasaste con la juerga, te has dejado 750 en una noche", 750, TipoSorpresa.PAGARCOBRAR));
        mazo.add(new Sorpresa("Te ha llegado un sobre con una tarjeta negra, eres un nuevo especulador", 3000, TipoSorpresa.CONVERTIRME));
//        Collections.shuffle(mazo);
    }

    private void inicializarJugadores(ArrayList<String> nombres) {
        nombres.forEach((n) -> {
            jugadores.add(new Jugador(n));
        });
    }

    private void inicializarTablero() {
        tablero = new Tablero();
    }

    public static Qytetet getInstance() {
        return instance;
    }

    public void siguienteJugador() {
        int jugadorSiguiente = (jugadores.indexOf(this.jugadorActual) + 1) % jugadores.size();
        jugadorActual = jugadores.get(jugadorSiguiente);
    }

    void salidaJugadores() {
        for (Jugador jugador : this.jugadores) {
            jugador.setCasillaActual(tablero.obtenerCasillaNumero(0));
            jugador.modificarSaldo(7500);
        }
//        int i = new Random().nextInt(jugadores.size());
//        jugadorActual = jugadores.get(i);
        jugadorActual = jugadores.get((int) (Math.random() * jugadores.size()));

    }

    public ArrayList<Casilla> propiedadesHipotecadasJugador(boolean hipotecadas) {
        ArrayList<Casilla> casillas = new ArrayList<>();
        for (TituloPropiedad propiedad : jugadorActual.obtenerPropiedadesHipotecadas(hipotecadas)) {
            casillas.add(propiedad.getCasilla());
        }
        return casillas;
    }

    public void inicicializarJuego(ArrayList<String> nombres) {
        inicializarJugadores(nombres);
        inicializarTablero();
        inicializarCartasSorpresa();
        salidaJugadores();
    }

    public boolean comprarTituloPropiedad() {
        boolean puedoComprar = jugadorActual.comprarTitulo();
        return puedoComprar;
    }

    public boolean edificarCasa(Casilla casilla) {
        boolean puedoEdificar = false;
        boolean sePuedeEdificar = false;
        if (casilla.soyEdificable()) {
            sePuedeEdificar = ((Calle)casilla).sePuedeEdificarCasa(jugadorActual.getFactorEspeculador());
            if (sePuedeEdificar) {
                puedoEdificar = jugadorActual.puedoEdificar(casilla);
                if (puedoEdificar) {
                    jugadorActual.modificarSaldo(-((Calle)casilla).edificarCasa());
                }
            }
        }
        return puedoEdificar;
    }

    public boolean edificarHotel(Casilla casilla) {
        boolean puedoEdificar = false;
        boolean sePuedeEdificar = false;
        if (casilla.soyEdificable()) {
            sePuedeEdificar = ((Calle)casilla).sePuedeEdificarHotel(jugadorActual.getFactorEspeculador());
            if (sePuedeEdificar) {
                puedoEdificar = jugadorActual.puedoEdificar(casilla);
                if (puedoEdificar) {
                    jugadorActual.modificarSaldo(-((Calle)casilla).edificarHotel());
                }
            }
        }
        return puedoEdificar;

    }

    public boolean hipotecarPropiedad(Casilla casilla) {
        boolean puedoHipotecar = false;
        if (((Calle)casilla).soyEdificable() && !((Calle)casilla).estaHipotecada()) {
            puedoHipotecar = jugadorActual.puedoHipotecar(casilla);
            if (puedoHipotecar) {
                jugadorActual.modificarSaldo(((Calle)casilla).hipotecar());
            }
        }
        return puedoHipotecar;
    }

    public boolean cancelarHipoteca(Casilla casilla) {
        boolean puedoCancelar = false;
        if (((Calle)casilla).estaHipotecada()) {
            puedoCancelar = jugadorActual.puedoHipotecar(casilla);
            if (puedoCancelar) {
                jugadorActual.modificarSaldo(-((Calle)casilla).cancelarHipoteca());
            }
        }
        return puedoCancelar;
    }

    public boolean venderPropiedad(Casilla casilla) {
        boolean puedoVender = false;
        if (casilla.soyEdificable()) {
            puedoVender = jugadorActual.puedoVenderPropiedad(casilla);
            if (puedoVender) {
                jugadorActual.venderPropiedad(casilla);
            }
        }
        return puedoVender;
    }

    public boolean aplicarSorpresa() {
        boolean tienePropietario = false;
        int valor = cartaActual.getValor();

        if (null != cartaActual.getTipo()) {
            switch (cartaActual.getTipo()) {
                case PAGARCOBRAR:
                    jugadorActual.modificarSaldo(valor);
                    break;
                case IRACASILLA:
                    boolean esCarcel = tablero.esCasillaCarcel(valor);
                    if (esCarcel) {
                        encarcelarJugador();
                    } else {
                        Casilla nuevaCasilla = tablero.obtenerCasillaNumero(valor);
                        tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
                    }
                    break;
                case PORCASAHOTEL:
                    jugadorActual.pagarCobrarPorCasaYHotel(valor);
                    break;
                case PORJUGADOR:
                    for (Jugador jugador : jugadores) {
                        if (jugador != jugadorActual) {
                            jugador.modificarSaldo(valor);
                            jugadorActual.modificarSaldo(-valor);
                        }
                    }
                    break;
                case CONVERTIRME:
                    int index = jugadores.indexOf(jugadorActual);
                    jugadorActual = jugadorActual.convertirme(valor);
                    jugadores.set(index, jugadorActual);
                    break;
                default:
                    break;
            }
        }

        if (cartaActual.getTipo() == TipoSorpresa.SALIRCARCEL) {
            jugadorActual.setCartaLibertad(cartaActual);
        } else {
            mazo.add(cartaActual);
        }

        return tienePropietario;
    }

    private void encarcelarJugador() {
        if (!jugadorActual.tengoCartaLibertad()) {
            jugadorActual.irACarcel(tablero.getCarcel());
        } else {
            mazo.add(jugadorActual.devolverCartaLibertad());
        }
    }

    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo) {
        boolean libre = false;
        if (metodo == MetodoSalirCarcel.TIRANDODADO) {
            valorDado = dado.tirar();
            libre = valorDado > 4;
        } else if (metodo == MetodoSalirCarcel.PAGANDOLIBERTAD) {
            libre = jugadorActual.pagarLibertad(PRECIO_LIBERTAD);
        }
        if (libre) {
            jugadorActual.setEncarcelado(false);
        }
        return libre;
    }

    public boolean jugar() {
        boolean tienePropietario = false;
        valorDado = dado.tirar();
        Casilla casillaPosicion = jugadorActual.getCasillaActual();
        Casilla nuevaCasilla = tablero.obtenerNuevaCasilla(casillaPosicion, 7);
        tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);

        if (!nuevaCasilla.soyEdificable()) {
            if (nuevaCasilla.getTipo() == TipoCasilla.JUEZ) {
                encarcelarJugador();
            } else if (nuevaCasilla.getTipo() == TipoCasilla.SORPRESA) {
                cartaActual = mazo.get(0);
                mazo.remove(0);
            }
        }
        return tienePropietario;
    }

    public Map obtenerRanking() {
        Map<String, Integer> ranking = new LinkedHashMap<>();
        ArrayList<Jugador> listaOrdenada = new ArrayList(jugadores);
        listaOrdenada.sort((p1, p2) -> p2.obtenerCapital() - p1.obtenerCapital());
        listaOrdenada.forEach((jugador)
                -> ranking.put(jugador.getNombre(), jugador.obtenerCapital()));
        return ranking;
    }
}
